### ISSUE TITLE
Failed to parse valid JSON

### ISSUE THREAD

**Issue by maedox** [Graylog2/graylog2-server#2048]
### Problem description
I have a bunch of errors in my server.log about problems parsing JSON (GELF input) from Apache2 log messages.

**Example 1:**
```
{ "version":"1.1", "timestamp":"1460373739.999", "host":"asset-mds", "short_message":"GET /api/rest/query/ExtendedMasterDataQuery HTTP/1.1", "level":"6", "_application_name":"apache2", "_balancer":"-", "_balancer_worker":"-", "_bytes_received":"410", "_bytes_sent":"1082", "_client_ip":"213.219.38.141", "_forensic_id":"-", "_forwarded_for":"-", "_http_host":"asset-mds.tracetracker.com", "_http_method":"GET", "_http_query":"?includeAttributes=true&includeChildren=true&vocabularyName=urn:tracetracker:1000002034:asset:mapping&EQATTR_device=hex:urn:imei:820898&pageSize=1000&page=1&", "_http_referer":"-", "_http_request_path":"/api/rest/query/ExtendedMasterDataQuery", "_http_status":"200", "_protocol":"HTTP/1.1", "_serve_time":"11409", "_tls_cipher":"ECDHE-RSA-AES128-SHA", "_tls_cipher_key_size":"128", "_tls_protocol":"TLSv1", "_user_agent":"TraceTracker/Asset" }
```

**The log says:**
```
2016-04-11T12:22:20.257+01:00 ERROR [DecodingProcessor] Unable to decode raw message a7c5ae72-ffd7-11e5-a86a-f23c91b06aae (journal offset 20905114) encoded as gelf received from unknown source.
2016-04-11T12:22:20.257+01:00 ERROR [DecodingProcessor] Error processing message RawMessage{id=a7c5ae72-ffd7-11e5-a86a-f23c91b06aae, journalOffset=20905114, codec=gelf, payloadSize=2048, timestamp=2016-04-11T11:22:20.247Z}
com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'tracker': was expecting 'null', 'true', 'false' or NaN
 at [Source: tracker:1000002034:asset:mapping&EQATTR_hex:urn:imei:820892=hex:urn:imei:820892&pageSize=1000&page=1&", "_http_referer":"-", "_http_request_path":"/api/rest/query/ExtendedMasterDataQuery", "_http_status":"200", "_protocol":"HTTP/1.1", "_serve_time":"10598", "_tls_cipher":"ECDHE-RSA-AES128-SHA", "_tls_cipher_key_size":"128", "_tls_protocol":"TLSv1", "_user_agent":"TraceTracker/Asset" }
```
JSONLint.com says it's valid.

**Example 2, from the log:**
```
2016-04-11T12:35:49.736+01:00 ERROR [DecodingProcessor] Unable to decode raw message 8a437150-ffd9-11e5-a86a-f23c91b06aae (journal offset 20924602) encoded as gelf received from unknown source.
2016-04-11T12:35:49.736+01:00 ERROR [DecodingProcessor] Error processing message RawMessage{id=8a437150-ffd9-11e5-a86a-f23c91b06aae, journalOffset=20924602, codec=gelf, payloadSize=626, timestamp=2016-04-11T11:35:49.733Z}
com.fasterxml.jackson.core.JsonParseException: Unrecognized character escape 'x' (code 120)
 at [Source: { "version":"1.1", "timestamp":"1460374549.727", "host":"asset-mds", "short_message":"\xc5\x06\xda\xab - HTTP/1.0", "level":"6", "_application_name":"apache2", "_balancer":"-", "_balancer_worker":"-", "_bytes_received":"5", "_bytes_sent":"0", "_client_ip":"169.229.3.91", "_forensic_id":"-", "_forwarded_for":"-", "_http_host":"ip-172-30-0-17.eu-west-1.compute.internal", "_http_method":"\xc5\x06\xda\xab", "_http_query":"", "_http_referer":"-", "_http_request_path":"-", "_http_status":"400", "_protocol":"HTTP/1.0", "_serve_time":"38", "_tls_cipher":"-", "_tls_cipher_key_size":"-", "_tls_protocol":"-", "_user_agent":"-" }
; line: 1, column: 89]
        at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:1581) ~[graylog.jar:?]
        at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:533) ~[graylog.jar:?]
        at com.fasterxml.jackson.core.base.ParserMinimalBase._handleUnrecognizedCharacterEscape(ParserMinimalBase.java:510) ~[graylog.jar:?]
        at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._decodeEscaped(ReaderBasedJsonParser.java:2208) ~[graylog.jar:?]
        at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._finishString2(ReaderBasedJsonParser.java:1773) ~[graylog.jar:?]
        at com.fasterxml.jackson.core.json.ReaderBasedJsonParser._finishString(ReaderBasedJsonParser.java:1747) ~[graylog.jar:?]
        at com.fasterxml.jackson.core.json.ReaderBasedJsonParser.getText(ReaderBasedJsonParser.java:233) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.BaseNodeDeserializer.deserializeObject(JsonNodeDeserializer.java:233) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.java:69) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.deser.std.JsonNodeDeserializer.deserialize(JsonNodeDeserializer.java:15) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3731) ~[graylog.jar:?]
        at com.fasterxml.jackson.databind.ObjectMapper.readTree(ObjectMapper.java:2272) ~[graylog.jar:?]
        at org.graylog2.inputs.codecs.GelfCodec.decode(GelfCodec.java:115) ~[graylog.jar:?]
        at org.graylog2.shared.buffers.processors.DecodingProcessor.processMessage(DecodingProcessor.java:136) ~[graylog.jar:?]
        at org.graylog2.shared.buffers.processors.DecodingProcessor.onEvent(DecodingProcessor.java:82) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:58) [graylog.jar:?]
        at org.graylog2.shared.buffers.processors.ProcessBufferProcessor.onEvent(ProcessBufferProcessor.java:35) [graylog.jar:?]
        at com.lmax.disruptor.WorkProcessor.run(WorkProcessor.java:139) [graylog.jar:?]
        at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [graylog.jar:?]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_77]
```

The last one I agree with, but how can I fix it? That's the message that apache2 sends, because some jackass tries some weird HTTP methods full of escape sequences. Any ideas?

**Example 3:**
```
2016-04-11T12:22:20.255+01:00 ERROR [DecodingProcessor] Unable to decode raw message a7c5ae71-ffd7-11e5-a86a-f23c91b06aae (journal offset 20905113) encoded as gelf received from unknown source.
2016-04-11T12:22:20.256+01:00 ERROR [DecodingProcessor] Error processing message RawMessage{id=a7c5ae71-ffd7-11e5-a86a-f23c91b06aae, journalOffset=20905113, codec=gelf, payloadSize=2048, timestamp=2016-04-11T11:22:20.247Z}
com.fasterxml.jackson.core.JsonParseException: Unexpected character ('"' (code 34)): Expected space separating root-level values
 at [Source: 8.141", "_forensic_id":"-", "_forwarded_for":"-", "_http_host":"asset-mds.tracetracker.com", "_http_method":"GET", "_http_query":"?includeAttributes=true&includeChildren=true&vocabularyName=urn:epcglobal:epcis:vtype:EPCClass&EQ_name=urn:gtnet:idpat:gsii:1000002025.12XXXX02.*&attributeNames=urn%3Atracetracker%3Aasset%3Amda%3Aasset_cat&ssoObject=EotvTrtnqKf5Uezxbol7vdcS7dJONhBEDuf3xA7bb%2B4%3D", "_http_referer":"-", "_http_request_path":"/mds/api/rest/query/SimpleMasterDataQuery", "_http_status":"200", "_protocol":"HTTP/1.1", "_serve_time":"16499", "_tls_cipher":"ECDHE-RSA-AES128-SHA", "_tls_cipher_key_size":"128", "_tls_protocol":"TLSv1", "_user_agent":"TraceTracker/Asset" }
```
As far as I can tell from the part that got logged is that it's valid JSON.

**Apache2 2.4.18 configuration (GELF):**
```
LogFormat "{ \"version\":\"1.1\",\
 \"timestamp\":\"%{sec}t.%{msec_frac}t\",\
 \"host\":\"${Hostname}\",\
 \"short_message\":\"%m %U %H\",\
 \"level\":\"6\",\
 \"_application_name\":\"apache2\",\
 \"_balancer\":\"%{BALANCER_NAME}e\",\
 \"_balancer_worker\":\"%{BALANCER_WORKER_NAME}e\",\
 \"_bytes_received\":\"%I\",\
 \"_bytes_sent\":\"%O\",\
 \"_client_ip\":\"%a\",\
 \"_forensic_id\":\"%{forensic-id}n\",\
 \"_forwarded_for\":\"%{X-Forwarded-For}i\",\
 \"_http_host\":\"%V\",\
 \"_http_method\":\"%m\",\
 \"_http_query\":\"%q\",\
 \"_http_referer\":\"%{Referer}i\",\
 \"_http_request_path\":\"%U\",\
 \"_http_status\":\"%>s\",\
 \"_protocol\":\"%H\",\
 \"_serve_time\":\"%D\",\
 \"_tls_cipher\":\"%{SSL_CIPHER}e\",\
 \"_tls_cipher_key_size\":\"%{SSL_CIPHER_USEKEYSIZE}e\",\
 \"_tls_protocol\":\"%{SSL_PROTOCOL}e\",\
 \"_user_agent\":\"%{User-agent}i\" }"\
 gelf

CustomLog "|/bin/nc -u graylog.example.com 12201" gelf
```

### Steps to reproduce the problem
Send in the logged JSON strings directly to Graylog I guess.

### Environment

* Graylog Version: 2.0 beta 2
* Elasticsearch Version:
* MongoDB Version:
* Operating System: Ubuntu 14.04
* Browser version:

**Comment by maedox** [comment]
For the example with control characters in the HTTP method, perhaps the parser should be told to allow them?

**Comment by joschi** [comment]
@maedox Your first example looks like valid JSON but could contain non-printable characters breaking the JSON payload.

The second example is (kind of) invalid JSON, using an escape code that's not covered by the specification:
* https://mathiasbynens.be/notes/javascript-escapes#single
* https://es5.github.io/x7.html#x7.8.4

Unfortunately the third example is incomplete, so I cannot say exactly why it failed.

FWIW, it's quite a gamble to simply put the logging values of Apache httpd into JSON without any escaping or sanitization, because those could easily contain invalid character sequences (e. g. simply contain a double quote) which could break the JSON payload.

**Comment by maedox** [comment]
Yes, that last one is incomplete. I think the Graylog logging has room for improvement. :wink:

Currently I don't see any good way to get logs from Apache2 to Graylog. You're right it is a gamble, but it mostly works well. There is a [GELF module](https://github.com/Graylog2/apache-mod_log_gelf) for Apache2, but I'm not sure I would trust that one either. It says beta, and it also requires me to use mpm_prefork.

**Comment by joschi** [comment]
@maedox While it's not the most elegant or efficient way to do this, you could keep writing the Apache httpd access logs to a file and pick that up via nxlog, filebeat, or logstash to send it to Graylog. At least the GELF messages will be valid that way.

The Apache httpd GELF logging module is still in beta, but you can of course help to improve it by using it and write bug reports. 😉

**Comment by coip** [comment]
For what it's worth, I've found that logs sent (at least with a rudimentary `curl -XPOST grayloghostname -d $payload`) with a field of 'timestamp' in $payload are dropped, though succeed when submit without the timestamp field. I just tested with the first half of your GELF'd data, and it wasn't found in my GL instance. Once I removed that field, it showed up successfully. Another aspect that was troubling at first is I was receiving a 202 back, without the timestamp'd log actually appearing. Can provide better documentation for replicating the event if this doesn't happen for you guys, @joschi. :-) As a work around, I have been prepending timestamp fields on the log-input side with an identifier to make it a unique field(avoiding collisions on GL side)
